lnrpc/routerpc/router_server.go

```diff
@@ -0,0 +1,221 @@
+// +build routerrpc
+
+package routerpc
+
+import (
+	"context"
+	fmt "fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+	"time"
+
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing"
+	"github.com/lightningnetwork/lnd/zpay32"
+	"google.golang.org/grpc"
+	"gopkg.in/macaroon-bakery.v2/bakery"
+)
+
+const (
+	// subServerName is the name of the sub rpc server. We'll use this name
+	// to register ourselves, and we also require that the main
+	// SubServerConfigDispatcher instance recognize as the name of our
+	subServerName = "Routerpc"
+)
+
+var (
+	// macaroonOps are the set of capabilities that our minted macaroon (if
+	// it doesn't already exist) will have.
+	macaroonOps = []bakery.Op{
+		{
+			Entity: "offchain",
+			Action: "write",
+		},
+	}
+
+	// macPermissions maps RPC calls to the permissions they require.
+	macPermissions = map[string][]bakery.Op{
+		"/routerpc.Router/SendPayment": {{
+			Entity: "offchain",
+			Action: "write",
+		}},
+	}
+
+	// DefaultRouterMacFilename is the default name of the router macaroon
+	// that we expect to find via a file handle within the main
+	// configuration file in this package.
+	DefaultRouterMacFilename = "router.macaroon"
+)
+
+// Server is a stand alone sub RPC server which exposes functionality that
+// allows clients to route arbitrary payment through the Lightning Network.
+type Server struct {
+	started  int32 // To be used atomically.
+	shutdown int32 // To be used atomically.
+
+	cfg *Config
+
+	quit chan struct{}
+}
+
+// A compile time check to ensure that Server fully implements the RouterServer
+// gRPC service.
+var _ RouterServer = (*Server)(nil)
+
+// fileExists reports whether the named file or directory exists.
+func fileExists(name string) bool {
+	if _, err := os.Stat(name); err != nil {
+		if os.IsNotExist(err) {
+			return false
+		}
+	}
+	return true
+}
+
+// New creates a new instance of the RouterServer given a configuration struct
+// that contains all external dependencies. If the target macaroon exists, and
+// we're unable to create it, then an error will be returned. We also return
+// the set of permissions that we require as a server. At the time of writing
+// of this documentation, this is the same macaroon as as the admin macaroon.
+func New(cfg *Config) (*Server, lnrpc.MacaroonPerms, error) {
+	// If the path of the router macaroon wasn't generated, then we'll
+	// assume that it's found at the default network directory.
+	if cfg.RouterMacPath == "" {
+		cfg.RouterMacPath = filepath.Join(
+			cfg.NetworkDir, DefaultRouterMacFilename,
+		)
+	}
+
+	// Now that we know the full path of the router macaroon, we can check
+	// to see if we need to create it or not.
+	macFilePath := cfg.RouterMacPath
+	if !fileExists(macFilePath) && cfg.MacService != nil {
```

Comment:
would be great if this logic were abstracted in `lnrpc`, so we don't have to reimplement it with each new subserver. i imagine all subservers we add will have to use macaroons anyway